[Policy] Several transaction standardness rules

jl2012 opened the pull request:
This disables `OP_CODESEPARATOR` in non-segwit scripts (even in an unexecuted branch), and makes a positive `FindAndDelete` result invalid. This ensures that the `scriptCode` serialized in `SignatureHash` is always the same as the script passing to the `EvalScript`.

jl2012 commented:
This is part of #8755
@TheBlueMatt

TheBlueMatt commented:
Strong Concept ACK. We do ML posts for new standardness rules, right? Probably doesn't matter given their lack of use.

sipa commented:
Any reason to support OP_CODESEPARATOR inside P2WSH?

TheBlueMatt commented:
I'd prefer we be consistent. Especially if OP_CODESEPARATOR has no meaning in SegWit transactions we should just make it nonstandard everywhere.

On September 29, 2017 6:18:36 PM EDT, Pieter Wuille <notifications@github.com> wrote:
>Any reason to support OP_CODESEPARATOR inside P2WSH?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/11423#issuecomment-333252262

jl2012 commented:
@sipa I think @NicolasDorier has something that use CODESEPARATOR. I think it could make the size of some contracts smaller.

This is actually part of the attempt to fix the pre-segwit quadratic sighash problem. A variable `scriptCode` makes sighash caching more difficult. That's why I proposed to do it for pre-segwit only

irieGhost commented:
<validation class="twitter-tweet"><p lang="en" dir="ltr">Port Townsend Terminal<br><br>1301 Water St, Port Townsend, WA 98368<br>+1 206-464-6400 <a href="https://t.co/3LEHrXpu7y">https://t.co/3LEHrXpu7y</a></p>&mdash; ahrlo jan ebio (@irieBrain) <a href="https://twitter.com/irieBrain/status/913990767797985280?ref_src=twsrc%5Etfw">September 30, 2017</a></blockquote> <script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>

NicolasDorier commented:
```
Any reason to support OP_CODESEPARATOR inside P2WSH?
```

YES, I am using it for tumblebit, it saves 33 bytes per outputs. It allows you to sign a specific branch.

NicolasDorier commented:
Just to explain my case here:
Imagine the following: Alice, expect Bob to publish a transaction satisfying the PAYMENT condition to get the preimage of <BOB_HASH>.

```
OP_DEPTH 3 EQUAL
OP_IF
    2 <ALICE_KEY_PAYMENT> <BOB_KEY_PAYMENT> MULTICHECKSIGVERIFY <BOB_HASH> HASH160 EQUAL
OP_ELSE
    <ALICE_KEY_REDEEM> CHECKSIG CLTV DROP
OP_END
```

If ALICE_KEY_PAYMENT == ALICE_KEY_REDEEM, this would be no good, as BOB could use Alice signature to satisfy the REDEEM condition, and get the money without revealing the pre image.

Now you can save 33 bytes this way. (untested, but it get the idea of what I do for TB)

```
PUSH <ALICE_KEY>
OP_DEPTH 3 EQUAL
OP_IF
    OP_SWAP <BOB_KEY_PAYMENT> CHECKSIGVERIFY OP_CODESEP <BOB_HASH> HASH160 EQUAL
OP_ELSE
     CLTV DROP
OP_END
CHECKSIG
```

Notifce the Alice can just pass a SIGHASH_NONE signature for the first branch to Bob, so she does not have to know how bob will spend the txout, while still being sure she can get the preimage.

TheBlueMatt commented on src/script/interpreter.h:
```diff
@@ -106,6 +106,10 @@ enum
     //
     SCRIPT_VERIFY_WITNESS_PUBKEYTYPE = (1U << 15),
+
+    // Making OP_CODESEPARATOR and FindAndDelete non-standard in non-segwit scripts
```
For correctness, may wish to note that if FindAndDelete matches in either SegWit or non-SegWit, the script fails, OP_CODESEPARATOR limitation only applies to non-SegWit.

jl2012 replied:
`FindAndDelete` is not used anywhere in segwit

TheBlueMatt replied:
Oops, indeed, missed the sigversion check.

TheBlueMatt replied:
nit: comments on script flags shouldnt reference standardness - this flag "Makes OP_CODESEPARATOR and FindAndDelete fail any non-segwit scripts", not just make them nonstandard.

NicolasDorier commented on src/script/interpreter.cpp:
```diff
@@ -301,6 +301,10 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 opcode == OP_RSHIFT)
                 return set_error(serror, SCRIPT_ERR_DISABLED_OPCODE); // Disabled opcodes.
 
+            // OP_CODESEPARATOR in non-segwit transaction is invalid even in an unexecuted branch
```
nit: would have pushed that in the switch case.

luke-jr replied:
The switch case wouldn't get reached if it's in an unexecuted branch...?

NicolasDorier replied:
Yep, disregard what I said, it would not have worked.

jgarzik commented:
concept ACK

jl2012 committed:
Add transaction tests for constant scriptCode

Tests showing that CONST_SCRIPTCODE is applied only to non-segwit transactions

TheBlueMatt reviewed (commented):
utACK 98cea79fd2161ce0b42bdef27befc6a23047975f

luke-jr commented on src/script/interpreter.cpp:
```diff
@@ -301,6 +301,10 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 opcode == OP_RSHIFT)
                 return set_error(serror, SCRIPT_ERR_DISABLED_OPCODE); // Disabled opcodes.
 
+            // OP_CODESEPARATOR in non-segwit transaction is invalid even in an unexecuted branch
```
"Invalid" is the wrong term here, since it's just a policy rule (for now).

luke-jr reviewed (commented):
Concept ACK

TheBlueMatt commented:
utACK effb6f9568a92ad6fe0ebf9da308cb0237df327b. One interesting test-case you may consider adding is checking that a FindAndDelete does not match (and thus the script is valid, even with CONST_SCRIPTCODE) a push with a non-minimal push encoding.

jonasschnelli commented:
utACK 0575b1831cd52987c76320d304674a27a140fe1f
* IMO requires bitcoin-dev ML post (if merged)
* requires release notes part (can be done after merging)

Sjors commented on src/script/interpreter.cpp:
```diff
@@ -301,6 +301,10 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 opcode == OP_RSHIFT)
                 return set_error(serror, SCRIPT_ERR_DISABLED_OPCODE); // Disabled opcodes.
 
+            // With SCRIPT_VERIFY_CONST_SCRIPTCODE, OP_CODESEPARATOR in non-segwit script is invalid even in an unexecuted branch
```
Perhaps out of scope, but the word "invalid" is confusing. It's non-standard because this check is only part of `STANDARD_SCRIPT_VERIFY_FLAGS`. A future soft fork could make it actually invalid depending on e.g. block height. Perhaps we should use a different word like "reject"?

TheBlueMatt replied:
I mean it clearly indicates that "With SCRIPT_VERIFY_CONST_SCRIPTCODE", and in that case it is "invalid", same as any other invalid action.c

esotericnonsense replied:
Not particularly concerned but I agree that 'rejected' would be slightly clearer here.

TheBlueMatt commented:
Relevant ML thread is at https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-November/015292.html

NicolasDorier commented:
Just to clarify, OP_CODESEP could be removed completely from segwit if there is another way to explicitely sign a script path. (MAST could allow that)

jl2012 commented:
rebased

esotericnonsense commented:
utACK 850c41c2a33efd73eff0bbdefc6ba2762901b60e
checked the diff of tx_valid and tx_invalid.json

jl2012 commented:
Rebase with comments fixed (s/invalid/rejected/)

jl2012 committed:
Policy to reject extremely small transactions

A transaction with 1 segwit input and 1 P2WPHK output has non-witness size of 82 bytes. Anything smaller than this have unnecessary malloc overhead and are not relayed/mined.

jl2012 committed:
Add constant scriptCode policy in non-segwit scripts

This disables OP_CODESEPARATOR in non-segwit scripts (even in an unexecuted branch), and makes a positive FindAndDelete result invalid. This ensures that the scriptCode serialized in SignatureHash() is always the same as the script passing to the EvalScript.

laanwj commented:
Rebased (for 6a7456ad6072f405e8b02bffa0fb4e9f0cfe71e0) version here: https://github.com/laanwj/bitcoin/tree/2018_05_temp_11423
Feel free to take over.

Commit:
qa: Pad scriptPubKeys to get minimum sized txs

NicolasDorier commented:
utACK.

TheBlueMatt commented:
utACK 364bae5f7a6b16eef63990154e48f19e7e693039

sipa commented:
utACK 364bae5f7a6b16eef63990154e48f19e7e693039

laanwj commented:
utACK 364bae5f7a6b16eef63990154e48f19e7e693039

fanquake commented:
Backported in #13317.
